public/app/features/alerting/unified/rule-list/hooks/useFilteredRulesIterator.ts

```diff
@@ -52,48 +48,113 @@ export function useFilteredRulesIteratorProvider() {
-      return from(generator).pipe(
-        map((group) => [ds, group] as const),
-        catchError(() => empty())
+    // If no data sources, just return Grafana rules
+    if (isEmpty(externalRulesSourcesToFetchFrom)) {
+      return grafanaRulesGenerator;
+    }
+
+    // Create a generator for each data source
+    const dataSourceGenerators = externalRulesSourcesToFetchFrom.map((dataSourceIdentifier) => {
+      const dsGenerator = filterDataSourceRules(
+        prometheusGroupsGenerator(dataSourceIdentifier, groupLimit),
+        dataSourceIdentifier,
+        normalizedFilterState
       );
+      return from(dsGenerator).pipe(catchError(() => empty()));
     });
 
-    // if we have no prometheus data sources, use an empty async iterable
-    const source = sourceIterables.at(0) ?? empty();
-    const otherIterables = sourceIterables.slice(1);
-
-    const dataSourcesIterator = merge(source, ...otherIterables).pipe(
-      filter(([_, group]) => groupFilter(group, normalizedFilterState)),
-      flatMap(([rulesSource, group]) => group.rules.map((rule) => [rulesSource, group, rule] as const)),
-      filter(([_, __, rule]) => ruleFilter(rule, filterState)),
-      map(([rulesSource, group, rule]) => mapRuleToRuleWithOrigin(rulesSource, group, rule))
-    );
-
-    return merge(grafanaIterator, dataSourcesIterator);
+    // Merge all generators
+    return merge(grafanaRulesGenerator, ...dataSourceGenerators);
   };
 
   return { getFilteredRulesIterator };
 }
 
+/**
+ * Flattens groups to rules and filters them
+ */
+async function* filterGrafanaRules(
+  groupsGenerator: AsyncIterable<GrafanaPromRuleGroupDTO[]>,
+  filterState: RulesFilter
+): AsyncGenerator<RuleWithOrigin, void, unknown> {
+  for await (const groups of groupsGenerator) {
+    for (const group of groups) {
+      if (!groupFilter(group, filterState)) {
+        continue;
+      }
+
+      for (const rule of group.rules) {
+        if (!ruleFilter(rule, filterState)) {
+          continue;
+        }
+
+        yield mapGrafanaRuleToRuleWithOrigin(group, rule);
+      }
+    }
+  }
+}
+
+/**
+ * Flattens groups to rules and filters them
+ */
+async function* filterDataSourceRules(
+  groupsGenerator: AsyncIterable<PromRuleGroupDTO[]>,
+  dataSourceIdentifier: DataSourceRulesSourceIdentifier,
+  filterState: RulesFilter
+): AsyncGenerator<RuleWithOrigin, void, unknown> {
+  for await (const groups of groupsGenerator) {
+    for (const group of groups) {
+      if (!groupFilter(group, filterState)) {
+        continue;
+      }
+
+      for (const rule of group.rules) {
+        if (!ruleFilter(rule, filterState)) {
+          continue;
+        }
+
+        yield mapRuleToRuleWithOrigin(dataSourceIdentifier, group, rule);
+      }
+    }
+  }
+}
+
+// find all data sources that the user might want to filter by, only allow Prometheus and Loki data source types
+const SUPPORTED_RULES_SOURCE_TYPES = ['loki', 'prometheus'];
```

Comment:
nit: hoist this somewhere in a constants file maybe?

Reply:
Fortunately, we already have such array defined in a different file 😃